### PR TITLE
Add RelayId constraint

### PR DESCRIPTION
**Pull request by cliedeman**
First attempt at adding a relay id constraint

TODO: update docs

### CODE REVIEW

**Comment by bbakerman** [comment]
I really appreciate the time take to contrib this to the library

However I am not sure its general purpose enough to warrant inclusion.

That is the use of `Relay.ResolvedGlobalId resolvedGlobalId = RELAY.fromGlobalId(value);`

This now says that you MUST use the graphql-java Relay class as its Identifier

Some might but others would want to use some other opaque id.  

This is a tricky problem because now we would need a plugin system to find the lowlevel "replay validation" code

Perhaps that not crazy - perhaps we can use Java ServiceLoader list style plugin patterns to find an implementation.

I mean I like all of the idea and 95% of this implementation but I worry that the specific format means it no great for all Relay users - that is the directive `@RelayID` will be used by this library that ties you too closely to a specific format 



https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html

**Comment by cliedeman** [comment]
Hi @bbakerman ,

Thanks for taking the time to look at this.

For me the main reason for this constraint is the types checking. E.g. Person:1. The reason is that during development this is an easy mistake to make, both when passing in values and returning them.  So this isn't actually meant to catch errors in production but provide faster feedback in development when an ID is not encoded.

For the general UUID id's this constraint really does not offer anything useful I think. This constraint really shines if the type is embedded in the ID.

Do you think changing isValidRelayIds to an interface and passing it in the constructor would be a valid approach?

Then by default we include the relay id encoding included in graphql-java and a user using a custom relay id encoding can provide a custom decoder?

`new RelayIdConstraint((globalId) => {some custom decoding logic that returns Relay.ResolvedGlobalId})`

Ciaran

**Comment by bbakerman** [comment]
On vacation right now... But mostly yes

**Comment by cliedeman** [comment]
@bbakerman [Directive wiring support](https://github.com/graphql-java-kickstart/graphql-spring-boot/commit/fc5fd02231142c278ea3b26dbd66d1c021064a9f) just got added to graphql-java-tools so my interest in this has been renewed. 

Are you still interested in merging this? Happy to make any changes

File: src/main/java/graphql/validation/constraints/standard/RelayIdConstraint.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                .applicableTypeNames(Scalars.GraphQLString.getName())
+
+                .directiveSDL("directive @RelayID(types : [String!] = [],  message : String = \"%s\" ) " +
+                        "on ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION", getMessageTemplate())
+                .build();
+    }
+
+    @Override
+    public boolean appliesToType(GraphQLInputType inputType) {
+        return isOneOfTheseTypes(inputType, Scalars.GraphQLID);
```
Comment (cliedeman) [inline src/main/java/graphql/validation/constraints/standard/RelayIdConstraint.java]: this should be
```isOneOfTheseTypes(Util.unwrapOneAndAllNonNull(inputType), Scalars.GraphQLID)```
to support `[ID!]` and `[ID!]!`

**Pull request closed without merging**